revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,232 @@
+// Package OCSP provides methods for checking the OCSP revocation status of a
+// certificate chain, as well as errors related to these checks
+package ocsp
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ocsp"
+)
+
+// Options specifies values that are needed to check OCSP revocation
+type Options struct {
+	CertChain   []*x509.Certificate
+	SigningTime time.Time
+	HTTPClient  *http.Client
+}
+
+const (
+	pkixNoCheckOID      string = "1.3.6.1.5.5.7.48.1.5"
+	ocspMaxResponseSize int64  = 20480 //bytes
+)
+
+// OCSPStatus checks OCSP based on the passed options and returns a 2D array of
+// errors. The size of these dimensions is as follows:
+//   - 1st dimension will always be equal in length to the certificate chain.
+//   - 2nd will have a length between 1 and the number of OCSPServers for the cert
+//
+// At each index, nil will be used to indicate no error. If there is an error,
+// it will be one of the errors defined in errors.go.
+//
+// (e.g. for a chain with 3 certs (leaf, intermediate, root) with both the leaf
+// and intermediate specifying 2 OCSP server, this could be a possible result):
+//
+//	[[PKIXNoCheckError, TimeoutError], [RevokedError], [nil]]
+func OCSPStatus(opts Options) [][]error {
+	certResults := make([][]error, len(opts.CertChain))
+
+	if len(opts.CertChain) == 0 {
+		return certResults
+	}
+
+	// Validate cert chain structure
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			if err := validateIssuer(cert, opts.CertChain[i+1]); err != nil {
+				fillResultsWithError(certResults, CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")})
+				return certResults
+			}
+		} else {
+			if err := validateIssuer(cert, cert); err != nil {
+				fillResultsWithError(certResults, CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")})
+				return certResults
+			}
+		}
+	}
+
+	// Check status for each cert in cert chain
+	var wg sync.WaitGroup
+	for i, cert := range opts.CertChain[:len(opts.CertChain)-1] {
+		wg.Add(1)
+		// Assume cert chain is accurate and next cert in chain is the issuer
+		go func(i int, cert *x509.Certificate) {
+			defer wg.Done()
+			certResults[i] = certOCSPStatus(cert, opts.CertChain[i+1], opts)
+		}(i, cert)
+	}
+	// Last is root cert, which will never be revoked by OCSP
+	certResults[len(opts.CertChain)-1] = []error{nil}
+
+	wg.Wait()
+	return certResults
+}
+
+func fillResultsWithError(results [][]error, err error) {
+	for i := range results {
+		results[i] = []error{err}
+	}
```

Comment:
This is only used for chain errors. The rationale is that it is an issue with the structure of the chain itself, it might not be certificate specific. One example might be if the chain is backwards. It isn't an issue with the first certificate, it is an issue with the chain itself. Hence every certificate gets the same error.

Reply:
This can be misleading. The root of the problem seems to be not having a first-class struct for this method's response. It seems complicated to convey the result (either success or failure) with this 2D array. Have you thought about replacing the 2D array with a better structure?